### PR TITLE
Add api docs

### PR DESCRIPTION
**Pull request by cldwalker**
This adds API docs for LightTable building off of https://github.com/LightTable/LightTable/commit/f94bf3af0d382424c69161d5b01680d0ea657a38 so we can encourage plugin authors and more users to read and understand the code base. I've included a script and documentation for publishing these docs. [Here is what the current docs look like](http://lighttable.github.io/LightTable/api/lt.macros.html). The styling isn't amazing but I'm ok with for a first pass. I did have to mark some fns private and added a few more docstring for public fns. I limited docs to a few namespaces as covered in #1955. I'm open to adding more down the road.

There weren't too many choices when choosing between libraries that generate documentation for cljs projects: I only found codox and marginalia. I chose codox because only it supported links to fns e.g. [object/add-tags](http://lighttable.github.io/LightTable/api/lt.object.html#var-add-tags) and I think it's important to encourage conversation around docs.

@rundis @kenny-evitt Would like to get this merged before our release

### CODE REVIEW

File: src/lt/macros.clj [hunk 1]
```diff
@@ -47,30 +47,30 @@
     [(first body) (rest body)]
     [[] body]))
 
-(defmacro on [name & body]
+(defmacro ^:private on [name & body]
```
Comment (rundis) [inline src/lt/macros.clj]: Unless you really think these should be private / tagged as private shouldn't you use the :no-doc true meta tag https://github.com/weavejester/codox#metadata-options?

Comment (cldwalker) [inline src/lt/macros.clj]: I am aware of :no-doc. My intent is to keep as much private as possible until we know we want them to be public

**Comment by rundis** [comment]
I found [codeina](https://github.com/funcool/codeina) to provide a lot nicer output than codox. Changing from codox to codeina shouldn't be hard they have very overlapping syntax/usage interface.
I've used it with pretty good success on my only cljs lib for general consumption [rewrite-cljs](http://rundis.github.io/rewrite-cljs/).

All though cljs doesn't honor any encapsulation with regards to private, I don't think the doc generation should be the deciding factor for what is considered private or not. Heck as of now the whole of the cljs code in LT could be considered it's public interface. Who knows what plugin authors might have started using. Ah anyway that's a entire discussion on it's own, 

I suggest to use the codox meta data tag :no-doc true, once that is fixed I'd say go for merge.

Btw I think this is a big step in the right direction of making LT more approachable for developers considering hacking away at their editor.

**Comment by cldwalker** [comment]
@rundis codeina looks nice! Do you want to try that in a PR after the release? Some things to about if we switch from codox to codeina: One advantage of codox is that weavejester has a longer track record of maintaining projects than funcool. One disadvantage of codeina is that they forked codox instead of working with them. To me that indicates a maintainer who gets excited in the short term and doesn't maintain their projects in the long term.

File: src/lt/macros.clj (hunk 1 above)
Comment (rundis) [inline src/lt/macros.clj]: Ok that makes sense. We probably should try to narrow down the public api quite a lot to make it maintainable (and a lot easier to make changes in the future without to much fear of breaking stuff).

**Comment by cldwalker** [comment]
@rundis Another option with codeina, is to take their styling and add it codox. weavejester [was very responsive today](https://github.com/weavejester/codox/issues/107) for an issue that I wanted to address around linking to source. Also, I think that is an issue we would have to solve in codeina

**Comment by rundis** [comment]
@cldwalker I guess codox is the closest thing to a defacto std for cljs docs so if we can help out improving that I'm all for that. Got you on the concern about track record and maintenance. Allthough I really hope funcool keeps it up, because we rely heavily on his buddy lib for our sso solution at work :-)

File: src/lt/object.cljs [hunk 2]
```diff
@@ -349,20 +351,20 @@
...
     inst))
 
-(defn refresh! [obj]
+(defn- refresh! [obj]
   (reset! obj (update-listeners obj))
   (raise* obj (trigger->behaviors :object.instant (:tags @obj)) nil)
   (raise obj :object.refresh))
 
-(defn add-behavior! [obj behavior]
+(defn- add-behavior! [obj behavior]
```
Comment (rundis) [inline src/lt/object.cljs]: This is used several places outside the namespace to dynamically add behavior (also I use it in one of my plugins to add test related behaviour after an editor has been opened). I guess for symmetry rem-behavior! should also be public.

File: src/lt/object.cljs [hunk 3]
```diff
@@ -40,33 +40,35 @@
...
-(defn ->triggers [behs]
+(defn- ->triggers [behs]
   (let [result (atom (transient {}))]
     (doseq [beh behs
             t (:triggers (->behavior beh))]
       (swap! result assoc! t (conj (or (get @result t) '[]) beh)))
     (persistent! @result)))
 
-(defn specificity-sort
+(defn- specificity-sort
```
Comment (rundis) [inline src/lt/object.cljs]: this is used in the keyboard ns (can't believe I remembered that !), but I'm not sure if the function belongs in the object namespace in the first place.

File: src/lt/object.cljs [hunk 4]
```diff
@@ -285,12 +287,12 @@
   [obj & r]
   (swap! obj #(apply update-in % r)))
 
-(defn assoc-in! [obj k v]
+(defn- assoc-in! [obj k v]
```
Comment (rundis) [inline src/lt/object.cljs]: I just used recently used this. There is a update! function (that is basically update-in, which would have been a better name btw), so assoc-in!, or assoc! would make sense for me to have at ones disposal.

File: src/lt/objs/editor.cljs [hunk 5]
```diff
@@ -125,23 +125,23 @@
   [e]
   (. (->cm-ed e) (getValue)))
 
-(defn ->token [e pos]
+(defn- ->token [e pos]
```
Comment (rundis) [inline src/lt/objs/editor.cljs]: I'm pretty sure this is used quite a bit (ref clojure plugin, and I have used it alot)  and needs to be public. Or perpaps there is a better way to get hold of the (cm) token at a given pos ?

File: src/lt/objs/editor.cljs [hunk 6]
```diff
@@ -125,23 +125,23 @@
...
   (js->clj (.getTokenAt (->cm-ed e) (clj->js pos)) :keywordize-keys true))
 
-(defn ->token-js [e pos]
+(defn- ->token-js [e pos]
   (.getTokenAt (->cm-ed e) (clj->js pos)))
 
-(defn ->token-type [e pos]
+(defn- ->token-type [e pos]
```
Comment (rundis) [inline src/lt/objs/editor.cljs]: I need this for 3 of my plugins !

File: src/lt/object.cljs (hunk 3 above)
Comment (cldwalker) [inline src/lt/object.cljs]: Agree its placement looked weird. Seems we should keep it private until we're sure of its purpose

File: src/lt/objs/editor.cljs (hunk 5 above)
Comment (cldwalker) [inline src/lt/objs/editor.cljs]: Not that I know of. Will change

File: src/lt/objs/editor.cljs (hunk 6 above)
Comment (cldwalker) [inline src/lt/objs/editor.cljs]: :) Will change

File: src/lt/object.cljs (hunk 4 above)
Comment (cldwalker) [inline src/lt/object.cljs]: Agree update-in is the better name. One day we should deprecate update!. Will change assoc-in!

File: src/lt/object.cljs (hunk 2 above)
Comment (cldwalker) [inline src/lt/object.cljs]: Will do

**Comment by cldwalker** [comment]
@rundis Made suggested changes. Anything else? I've made #2060, #2061 and #2062 which I think captures followup todos to this PR

**Comment by cldwalker** [comment]
@rundis I didn't make an issue for making more fns public for the 7 namespaces. Feel free to make one if there are more you'd like to consider

**Comment by kenny-evitt** [comment]
All of this looks good.

I concur with @cldwalker that it would probably be best to work with Codox instead of Codeina.

As for Marginalia, I'm confused what you meant that it doesn't 'supports links to fns'. AFAIK, the documentation is interleaved with the code so that's not an issue unless you were referring to, e.g. linking to function `bar` from the doc-string for function `foo`.

**Comment by cldwalker** [comment]
> As for Marginalia, I'm confused what you meant that it doesn't 'supports links to fns'. AFAIK, the documentation is interleaved with the code so that's not an issue unless you were referring to, e.g. linking to function bar from the doc-string for function foo

I meant the ability to have an external url to any var/fn e.g. http://lighttable.github.io/LightTable/api/lt.objs.files.html#var-copy. I don't think marginalia has that but perhaps I missed it

**Comment by rundis** [comment]
Looks good !

**Pull request merged**